ochre/Equipment/PV.py

```diff
@@ -1,71 +1,148 @@
+    # dc = pd.Series(system_model.Outputs.dc, index=time) / 1000  # in kW
+    return ac
 
 
 class PV(ScheduledLoad):
     """
-    PV System implemented using SAM or from an external schedule file.
+    PV System implemented using SAM or from an external schedule.
 
-    If using SAM, the PV capacity must be specified. Tilt and orientation can be specified, but will default to the
-    roof pitch and south.
+    If using SAM, the PV capacity must be specified. Tilt and azimuth can be specified, but will default to the
+    angle of the most southern facing roof.
 
-    If using an external schedule, the external schedule file must be specified.
+    If not using SAM, an external schedule must be specified as a DataFrame (via `schedule`) or as a file (as
+    `equipment_schedule_file`)
     """
     name = 'PV'
     end_use = 'PV'
     zone_name = None
 
-    def __init__(self, use_sam=None, capacity=None, tilt=None, orientation=None, include_inverter=True, **kwargs):
-        if use_sam is None:
-            use_sam = 'equipment_schedule_file' not in kwargs
-
-        if use_sam:
-            # Create PV schedule using SAM - requires capacity, tilt, orientation, and inverter efficiency
-            if capacity is None:
-                raise Exception('Must specify {} capacity (in kW) when using SAM'.format(self.name))
-            if tilt is None:
-                tilt = kwargs['roof pitch']
-            if orientation is None:
-                orientation = kwargs.get('building orientation', 0) % 360
-                if 90 < orientation <= 270:
-                    # use back roof when closer to due south (orientation always within 90 deg of south)
-                    orientation = (orientation + 180) % 360
-
-            inverter_efficiency = kwargs.get('inverter_efficiency') if include_inverter else 100  # in %
-            schedule = run_sam(tilt=tilt, orientation=orientation, inv_efficiency=inverter_efficiency, **kwargs)
+    def __init__(self,
+                 capacity=None,
+                 tilt=None,
+                 azimuth=None,
+                 envelope_model=None,
+                 inverter_capacity=None,
+                 inverter_efficiency=None,
+                 inverter_priority='Var',
+                 inverter_min_pf=0.8,
+                 **kwargs):
+        self.capacity = capacity  # in kW, DC
+        self.tilt = tilt
+        self.azimuth = azimuth
+
+        # get tilt and azimuth inputs to run SAM
+        if (tilt is None or azimuth is None):
+            if envelope_model is None:
+                raise OCHREException('Must specify PV tilt and azimuth, or provide an envelope_model with a roof.')
+            roofs = [bd.ext_surface for bd in envelope_model.boundaries if 'Roof' in bd.name]
+            # Use roof closest to south with preference to west (0-45 degrees)
+            roof_data = pd.DataFrame([[bd.tilt, az] for bd in roofs for az in bd.azimuths], columns=['Tilt', 'Az'])
```

Comment:
Have we checked this would work for a flat roof (say multifamily)?

Reply:
It should work for flat roofs, we should test that. It won't work for units
that have another unit above them. It's also not too hard to just specify the
tilt and azimuth in the `dwelling_args`.

Reply:
Just added an error. I think some BEopt tests would be good.